### PR TITLE
Nulog model training job

### PR DESCRIPTION
**Pull request by AmartC**
Commit contains the code used to train the Nulog model from logs obtained through Elasticsearch.

### CODE REVIEW

**Comment by tybalex** [comment]
why a lot of work shows up here although  they've already been merged to v0.0?

File: src/nulog-train/README.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+### Build image
```
Comment (sanjay920) [inline src/nulog-train/README.md:1]: Add a description of what the service is doing

File: src/nulog-train/nulog_train.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  name: nulog-train
+spec:
+  template:
+    metadata:
+      labels:
+        app: nulog-train
+    spec:
+      containers:
+      - name: nulog-train
+        image: amartyarancher/nulog-train
```
Comment (sanjay920) [inline src/nulog-train/nulog_train.yaml:13]: add specific tag `v0.0` to `amartyarancher/nulog-train`... should be `amartyarancher/nulog-train:v0.0`

File: src/nulog-train/train.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+)
+logging.info("Connected to Minio client")
+minio_client.meta.client.download_file(
+    "training-logs", "windows.tar.gz", "windows.tar.gz"
+)
+logging.info("Downloaded the windows.tar.gz file")
+shutil.unpack_archive("windows.tar.gz", format="gztar")
+logging.info("Extracted windows directory from windows.tar.gz file")
+
+WINDOWS_FOLDER_PATH = "windows/"
```
Comment (sanjay920) [inline src/nulog-train/train.py:34]: why is all this being run outside a function?

File: src/nulog-train/train.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+
+def train_nulog_model(minio_client):
+    log_format = "<Content>"
+    filters = '([ |:|\(|\)|\[|\]|\{|\}|"|,|=])'
```
Comment (sanjay920) [inline src/nulog-train/train.py:39]: why are filters here? preprocessing service takes care of all this

Comment (sanjay920) [reply src/nulog-train/train.py:39]: @tybalex please make sure train and inference code are aligned

File: src/nulog-train/README.md [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+```
+
+### Install NVIDIA GPU driver
+```
+kubectl apply -f https://raw.githubusercontent.com/NVIDIA/k8s-device-plugin/v0.6.0/nvidia-device-plugin.yml
+```
+
+### Start service
+```
+kubectl apply -f nulog_train.yaml
```
Comment (sanjay920) [inline src/nulog-train/README.md:13]: this is a `job` not a service. don't want to encourage the user to start this by themselves?

File: src/nulog-train/README.md [hunk 6]
```diff
@@ -0,0 +1,14 @@
+### Build image
+```
+docker build -t nulog-train ./
+```
+
+### Install NVIDIA GPU driver
+```
+kubectl apply -f https://raw.githubusercontent.com/NVIDIA/k8s-device-plugin/v0.6.0/nvidia-device-plugin.yml
```
Comment (sanjay920) [inline src/nulog-train/README.md:8]: should mention this is a pre-req

**Review by sanjay920 — CHANGES_REQUESTED** [review]
lot's of logging statements. remove if necessary. 
@AmartC how do you suggest we test this job? if you can provide some documentation on how to run this job in a silo it'd be great

**Comment by sanjay920** [comment]
> why a lot of work shows up here although they've already been merged to v0.0?

@AmartC is this addressed?

File: src/nulog-train/train.py (hunk 4 above)
Comment (tybalex) [reply src/nulog-train/train.py:39]: > @tybalex please make sure train and inference code are aligned

they are aligned. But yes the filters here can be removed as it not being used anymore.

Comment (tybalex) [reply src/nulog-train/train.py]: fixed.

File: src/nulog-train/train.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            aws_secret_access_key=MINIO_SECRET_KEY,
+            config=Config(signature_version="s3v4"),
+        )
+        logging.info("Connected to Minio client")
+        minio_client.meta.client.download_file(
+            "training-logs", "windows.tar.gz", "windows.tar.gz"
+        )
+        logging.info("Downloaded the windows.tar.gz file")
+        shutil.unpack_archive("windows.tar.gz", format="gztar")
+        logging.info("Extracted windows directory from windows.tar.gz file")
```
Comment (tybalex) [inline src/nulog-train/train.py:78]: probably not a good idea to have one `logging.info()` after each function running in this script. redundant Loggings should be removed.

File: src/nulog-train/train.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        "Published to model_ready Nats subject that new Nulog model is ready to be used for inferencing."
+    )
+    await nw.nc.close()
+
+
+if __name__ == "__main__":
+    try:
+        minio_client = boto3.resource(
+            "s3",
+            endpoint_url="http://minio.default.svc.cluster.local:9000",
```
Comment (tybalex) [inline src/nulog-train/train.py:67]: this can be an ENV variable?

File: src/nulog-train/train.py (hunk 7 above)
Comment (AmartC) [reply src/nulog-train/train.py:78]: Okay, I can remove these two logging statements at lines 76 and 78 but the other ones are good to have. Just give some context for where we are in the execution of the code.

File: src/nulog-train/train.py (hunk 8 above)
Comment (AmartC) [reply src/nulog-train/train.py:67]: yeah, agreed. Will change that.

**Review by tybalex — APPROVED** [review]
generally look good. one question I have: do we need the yaml file `nulog_train.yaml`? does the training controller launch training by applying this yaml file?

**Comment by AmartC** [comment]
Removed the yaml file as the way this job is called is through the training-controller service.

**Comment by AmartC** [comment]
@sanjay920 , yes the change was addressed.

**Comment by AmartC** [comment]
Closing this PR and going to resubmit from the forked Opni repository.

**Pull request closed without merging**